File: examples/c/app-char.c
#include <stdio.h>	/* printf */
#include <stdlib.h>	/* malloc, atoi, rand... */
#include <string.h>	/* memcpy, strlen... */
#include <stdint.h>	/* uints types */
#include <sys/types.h>	/* size_t ,ssize_t, off_t... */
#include <unistd.h>	/* close() read() write() */
#include <fcntl.h>	/* open() */
#include <sys/ioctl.h>	/* ioctl() */
#include <errno.h>	/* error codes */

#include "display.h"

int main(int argc, char** argv)
{
	int fd, len, retval;
	char buf[255];
	char cmd;

	if (argc < 2) {
		printf("Syntax: %s <device file path>\n", argv[0]);
		return -EINVAL;
	}

	if ((fd = open(argv[1], O_RDWR)) < 0) {
		fprintf(stderr, "Error opening file %s\n", argv[1]);
		return -EBUSY;
	}
	printf("file: %s opened!\n", argv[1]);

	while (1) {
		printf("Read or Write? [r,w]\nClose file [c]\n");
		scanf("%c%*c", &cmd);
		printf("CMD: %c\n", cmd);

		switch (cmd) {
		case 'r':
			printf("how many bytes you want to read?\n");
			scanf("%d%*c", &len);
			retval = read(fd, buf, len);
			buf[retval] = '\0';
			printf("red: %s. with %d bytes\n", buf, retval);
			buf[0] = '\0';
			break;
		case 'w':
			printf("type in what you want to write:\n");
			scanf("%[^\n]%*c", buf);
			retval = write(fd, buf, strlen(buf));
			printf("wrote %d bytes\n", retval);
			break;
		case 'c':
			close(fd);
			printf("file closed, exiting...\n");
			return EXIT_SUCCESS;
			break;
		default:
			printf("invalid input. try again\n");
			break;
		}
	}
}

File: examples/python/app-char.py
#!/usr/bin/python3

import sys

if len(sys.argv) < 2:
    print("Error: expected more command line arguments")
    print("syntax: %s </dev/device_file>"%sys.argv[0])
    exit(1)

while True:
    cmd = input("Open %s for Read or Write? [r/w]\nleave the script: [q]\n"%sys.argv[1])

    if cmd == 'r':
        fd = open(sys.argv[1], 'r')
        print("%s opened for read"%sys.argv[1])
        while True:
            cmd = input("How many bytes you want to read?\nclose the file: [c]\n")
            if cmd == 'c':
                fd.close()
                break
            else:
                retval = fd.read(int(cmd))
                print("red: %s"%retval)
    
    elif cmd == 'w':
        fd = open(sys.argv[1], 'w')
        print("%s opened for write"%sys.argv[1])
        while True:
            cmd = input("Type in what you want to write:\nclose the file: [c]\n")
            if cmd == 'c':
                fd.close()
                break
            else:
                retval = fd.write(cmd)
                print("wrote %s bytes"%int(retval))
    
    elif cmd == 'q':
        exit(0)
    
    else:
        print("input error")

